### PR TITLE
fix: ensure bounds vmaps as best as possible, more equinox errors, types for `__eq__`, `_BoundsD/I`, and `_PositionD/I`

### LINKED ISSUE #237: better runtime checking of inits
We should probably use equinox to do better runtime checking of values. For example, if you seed a Poisson deviate with a negative mean, the RNG returns -1 with every call.

```
In [6]: pd()
Out[6]: Array(-1., dtype=float32)

In [7]: pd()
Out[7]: Array(-1., dtype=float32)

In [8]: pd()
Out[8]: Array(-1., dtype=float32)

In [9]: pd()
Out[9]: Array(-1., dtype=float32)
```

### LINKED ISSUE #245: implement `_BoundsI` and `_BoundsD`
As it says on the tin...

### PR DESCRIPTION
**Pull request by beckermr**
This PR fixes a few other things with the bounds classes, images and typing.

- [x] ensure `BoundsD/I` vmaps as best as possible
- [x] use equinox to raise errors in the image classes
- [x] ensure all `__eq__` methods return JAX boolean values except when no values from the class are ever traced (e.g., two static `BoundsI` classes, interpolants)
- [x] docs
- [x] implement `_BoundsD/I`
- [x] implement `_PositionD/I`
- [x] add tests
  - [x] bool `__eq__` API
  - [x] bounds vmap
  - [x] bounds set
  - [x] bounds includes

closes #245 
closes #237

### CODE REVIEW

File: jax_galsim/bounds.py [hunk 1]
```diff
@@ -395,11 +374,176 @@ def to_galsim(self):
...
+        # bounds shape to static ints
+        ret = cls(
+            xmin=xmin,
+            deltax=xmax - xmin + 1,
+            ymin=ymin,
+            deltay=ymax - ymin + 1,
+        )
+        is_defined = jnp.where(
+            jnp.array(self._isdefined),
+            jnp.array(ret.deltax >= 0) & jnp.array(ret.deltay >= 0),
```
Comment (beckermr) [inline jax_galsim/bounds.py]:
```suggestion
            jnp.array(ret.deltax >= 1) & jnp.array(ret.deltay >= 1),
```

File: tests/jax/test_api.py [hunk 2]
```diff
@@ -360,6 +360,7 @@ def _reg_fun(p):
                         "xmax",
                         "ymax",
                         "isStatic",
+                        "isStaticShape",
```
Comment (beckermr) [inline tests/jax/test_api.py]:
```suggestion
```

File: jax_galsim/bounds.py [hunk 3]
```diff
@@ -596,10 +730,10 @@ def ymax(self, value):
 
     def _area(self):
         # Remember the + 1 this time to include the pixels on both edges of the bounds.
-        if not self.isDefined():
-            return 0
-        else:
+        if self._isdefined:
             return self.deltax * self.deltay
+        else:
+            return 0
```
Comment (beckermr) [inline jax_galsim/bounds.py]:
```suggestion
        if not self._isdefined:
            return 0
        else:
            return self.deltax * self.deltay
```

File: jax_galsim/bounds.py [hunk 4]
```diff
@@ -166,18 +153,32 @@ def origin(self):
...
         return self._center
 
     @property
     @implements(_galsim.Bounds.true_center)
     def true_center(self):
-        if not self.isDefined():
-            raise _galsim.GalSimUndefinedBoundsError(
-                "true_center is invalid for an undefined Bounds"
+        if not isinstance(self._isdefined, jnp.ndarray):
+            if not self.isDefined():
```
Comment (beckermr) [inline jax_galsim/bounds.py]:
```suggestion
            if not self._isdefined:
```

File: jax_galsim/bounds.py [hunk 5]
```diff
@@ -166,18 +153,32 @@ def origin(self):
     @property
     @implements(_galsim.Bounds.center)
     def center(self):
-        if not self.isDefined():
-            raise _galsim.GalSimUndefinedBoundsError(
-                "center is invalid for an undefined Bounds"
+        if not isinstance(self._isdefined, jnp.ndarray):
+            if not self.isDefined():
```
Comment (beckermr) [inline jax_galsim/bounds.py]:
```suggestion
            if not self._isdefined:
```

File: jax_galsim/image.py [hunk 6]
```diff
@@ -847,25 +859,30 @@ def calculate_fft(self):
...
-            max(
-                -self.bounds.xmin,
-                self.bounds.xmax + 1,
-            ),
-            max(
-                -self.bounds.ymin,
-                self.bounds.ymax + 1,
-            ),
-        )
+        # This has to be a static known constant since it is an array size
```
Comment (beckermr) [inline jax_galsim/image.py:850]:
```suggestion
        # TODO: figure out how to do FFT at fixed size and then reconstruct
        # the result. - MRB
        # This has to be a static known constant since it is an array size
```

File: jax_galsim/image.py [hunk 7]
```diff
@@ -280,6 +280,7 @@ def _get_xmin_ymin(array, kwargs, check_bounds=True):
             b = kwargs.pop("bounds")
             if not isinstance(b, BoundsI):
                 raise TypeError("bounds must be a galsim.BoundsI instance")
+
```
Comment (beckermr) [inline jax_galsim/image.py]:
```suggestion
```

File: jax_galsim/image.py [hunk 8]
```diff
@@ -571,6 +572,7 @@ def resize(self, bounds, wcs=None):
             raise GalSimImmutableError("Cannot modify an immutable Image", self)
         if not isinstance(bounds, BoundsI):
             raise TypeError("bounds must be a galsim.BoundsI instance")
+
```
Comment (beckermr) [inline jax_galsim/image.py]:
```suggestion
```

File: jax_galsim/image.py [hunk 9]
```diff
@@ -580,18 +582,17 @@ def resize(self, bounds, wcs=None):
     def subImage(self, bounds):
         if not isinstance(bounds, BoundsI):
             raise TypeError("bounds must be a galsim.BoundsI instance")
+
```
Comment (beckermr) [inline jax_galsim/image.py]:
```suggestion
```

File: jax_galsim/image.py [hunk 10]
```diff
@@ -619,18 +620,19 @@ def setSubImage(self, bounds, rhs):
             raise GalSimImmutableError("Cannot modify an immutable Image", self)
         if not isinstance(bounds, BoundsI):
             raise TypeError("bounds must be a galsim.BoundsI instance")
+
```
Comment (beckermr) [inline jax_galsim/image.py]:
```suggestion
```

File: jax_galsim/image.py [hunk 11]
```diff
@@ -848,24 +846,31 @@ def calculate_fft(self):
...
-        if self.bounds == full_bounds:
-            # Then the image is already in the shape we need.
+        if (
+            self.bounds.deltax == full_bounds.deltax
+            and self.bounds.deltay == full_bounds.deltay
+        ):
             ximage = self
         else:
-            # Then we pad out with zeros
             ximage = Image(full_bounds, dtype=self.dtype, init_value=0)
```
Comment (beckermr) [inline jax_galsim/image.py:873]:
```suggestion
            # Then we pad out with zeros
            ximage = Image(full_bounds, dtype=self.dtype, init_value=0)
```

File: jax_galsim/image.py [hunk 12]
```diff
@@ -848,21 +846,29 @@ def calculate_fft(self):
...
+                No2 = int(No2.item())
 
         full_bounds = BoundsI(xmin=-No2, deltax=2 * No2, ymin=-No2, deltay=2 * No2)
-        if self.bounds == full_bounds:
-            # Then the image is already in the shape we need.
+        if (
+            self.bounds.deltax == full_bounds.deltax
+            and self.bounds.deltay == full_bounds.deltay
+        ):
             ximage = self
```
Comment (beckermr) [inline jax_galsim/image.py:870]:
```suggestion
            # Then the image is already in the shape we need.
            ximage = self
```

File: jax_galsim/image.py [hunk 13]
```diff
@@ -902,21 +909,30 @@ def calculate_inverse_fft(self):
...
+                No2 = int(No2.item())
+
         target_bounds = BoundsI(xmin=0, deltax=No2 + 1, ymin=-No2, deltay=2 * No2)
-        if self.bounds == target_bounds:
-            # Then the image is already in the shape we need.
+        if (
+            self.bounds.deltax == target_bounds.deltax
+            and self.bounds.deltay == target_bounds.deltay
+        ):
             kimage = self
```
Comment (beckermr) [inline jax_galsim/image.py:931]:
```suggestion
            # Then the image is already in the shape we need.
            kimage = self
```

File: jax_galsim/interpolatedimage.py [hunk 14]
```diff
@@ -328,17 +328,24 @@ def __str__(self):
...
-            and self._stepk == other._stepk
-            and self._maxk == other._maxk
-        )
+        if self is other:
+            return jnp.array(True)
+        elif isinstance(other, InterpolatedImage):
+            return (
+                (self._xim == other._xim)
+                & (self.x_interpolant == other.x_interpolant)
+                & (self.k_interpolant == other.k_interpolant)
```
Comment (beckermr) [inline jax_galsim/interpolatedimage.py]:
```suggestion
                & jnp.array(self.x_interpolant == other.x_interpolant)
                & jnp.array(self.k_interpolant == other.k_interpolant)
```

File: jax_galsim/wcs.py
Comment (beckermr) [inline jax_galsim/wcs.py]:
```suggestion
```

File: tests/jax/test_api.py [hunk 15]
```diff
@@ -152,6 +170,16 @@ def _run_object_checks(obj, cls, kind):
 
         # check that we cannot hash the object
         hash(obj)
+
+        # check that val jax array
+        if (
+            hasattr(obj, "isStatic")
+            and obj.isStatic()
```
Comment (beckermr) [inline tests/jax/test_api.py]:
```suggestion
            (hasattr(obj, "isStatic")
            and obj.isStatic())
```

File: tests/jax/test_api.py [hunk 16]
```diff
@@ -126,6 +126,16 @@ def _run_object_checks(obj, cls, kind):
 
         # check that we can hash the object
         hash(obj)
+
+        # check that val jax array
+        if (
+            hasattr(obj, "isStatic")
+            and obj.isStatic()
```
Comment (beckermr) [inline tests/jax/test_api.py]:
```suggestion
            (hasattr(obj, "isStatic")
            and obj.isStatic())
```

File: docs/sharp-bits.rst [hunk 17]
```diff
@@ -89,6 +102,24 @@ These rules allow JAX-GalSim to transparently handle JAX's tracing operations, b
 the code raising generic ``Exception`` instances instead of more specific ``GalSim`` exceptions in
 some cases.
 
+Object Comparison with the ``==`` Operator
+------------------------------------------
+
+In JAX-GalSim, all objects which define arrays to be traced by JAX will return JAX boolean
+array scalars (i.e., ``jax.numpy.array(True)`` or ``jax.numpy.array(False)``) as the result
+of the ``==`` operator, otherwise they return Python boolean values. Important cases of this
+rule are static ``BoundsI`` objects and ``Interpolant`` objects (and their subclasses), which
```
Comment (beckermr) [inline docs/sharp-bits.rst]:
```suggestion
rule are static ``BoundsI`` objects, ``Interpolant`` objects (and their subclasses), and ``GSParams`` objects, all of which
```

File: docs/sharp-bits.rst [hunk 18]
```diff
@@ -89,6 +102,24 @@ These rules allow JAX-GalSim to transparently handle JAX's tracing operations, b
 the code raising generic ``Exception`` instances instead of more specific ``GalSim`` exceptions in
 some cases.
 
+Object Comparison with the ``==`` Operator
+------------------------------------------
+
+In JAX-GalSim, all objects which define arrays to be traced by JAX will return JAX boolean
+array scalars (i.e., ``jax.numpy.array(True)`` or ``jax.numpy.array(False)``) as the result
+of the ``==`` operator, otherwise they return Python boolean values. Important cases of this
```
Comment (beckermr) [inline docs/sharp-bits.rst]:
```suggestion
of the ``==`` operator. Otherwise the return value is a Python boolean. Important cases of this
```

File: tests/jax/test_api.py [hunk 19]
```diff
@@ -355,20 +379,24 @@ def _reg_fun(p):
...
+                    # jax-galsim BoundsI classes do not store xmin, ymin
+                    if issubclass(cls, jax_galsim.BoundsI) and method in [
+                        "xmin",
+                        "ymin",
+                        "deltax",
+                        "deltay",
+                    ]:
+                        continue
+
+                    # jax-galsim Bounds classes do not store xmin, ymin
```
Comment (beckermr) [inline tests/jax/test_api.py]:
```suggestion
                    # jax-galsim Bounds classes do not store xmax, ymax
```

File: tests/jax/test_api.py [hunk 20]
```diff
@@ -355,20 +379,24 @@ def _reg_fun(p):
                     ):
                         continue
 
+                    # jax-galsim BoundsI classes do not store xmin, ymin
```
Comment (beckermr) [inline tests/jax/test_api.py:390]:
```suggestion
                    # jax-galsim BoundsI classes do not store xmin, ymin
                    # or deltax/y directly
```

**Comment by beckermr** [comment]
OK @ismael-mendoza this one is ready for review. I hope this is the last major bit of work on these classes for a long time!

**Comment by ismael-mendoza** [comment]
Thanks Matt! Will start taking a look today

File: jax_galsim/bounds.py [hunk 21]
```diff
@@ -545,58 +729,118 @@ def _check_scalar(self, x, name):
...
+        # though it is an int so that autodiff works properly (needs floats in general).
+        # thus we cast here.
+        return cast_to_int(self._xmin)
 
     @xmin.setter
     def xmin(self, value):
+        value = check_is_int_then_cast(value, "BoundsI xmin values must be integers")
         if self._isstatic:
+            if self._dotypeconversion:
+                # attempt to convert widths to static values
```
Comment (ismael-mendoza) [inline jax_galsim/bounds.py]:
```suggestion
                # attempt to convert to static values
```

File: jax_galsim/bounds.py [hunk 22]
```diff
@@ -545,58 +729,118 @@ def _check_scalar(self, x, name):
...
+        # though it is an int so that autodiff works properly (needs floats in general).
+        # thus we cast here.
+        return cast_to_int(self._ymin)
 
     @ymin.setter
     def ymin(self, value):
+        value = check_is_int_then_cast(value, "BoundsI ymin values must be integers")
         if self._isstatic:
+            if self._dotypeconversion:
+                # attempt to convert widths to static values
```
Comment (ismael-mendoza) [inline jax_galsim/bounds.py]:
```suggestion
                # attempt to convert to static values
```

File: jax_galsim/bounds.py [hunk 23]
```diff
@@ -14,26 +15,33 @@
...
+  This restriction mirrors the JAX restriction that arrays have fixed shapes when traced
+  for function transformations like ``jax.vmap``, ``jax.jit``, etc.
+- Upon initialization, if a ``BoundsI`` object has a non-static shape, JAX-GalSim will attempt to convert
+  it to a static shape by extracting the dimensions from the array via ``.item()``. This operation will
+  cause JAX to raise an error if the code is being traced. JAX-Galsim performs the same conversion operation
+  when the ``deltax`` or ``deltay`` properties are set to non-static values via assignment.
+- If a ``BoundsI`` object is declared with static ``xmin`` and ``ymin`` values, and then one attempts to
+  convert them to non-static values via assignment, JAX-GalSim will attempt to convert the assigned values
+  back to static values. This operation will raise an error if the code is being traced.
+- ``Bounds`` classes in JAX-GalSim have an etxra method, ``isStatic`` that returns ``True`` if the object
```
Comment (ismael-mendoza) [inline jax_galsim/bounds.py]:
```suggestion
- ``Bounds`` classes in JAX-GalSim have an extra method, ``isStatic`` that returns ``True`` if the object
```

File: jax_galsim/bounds.py [hunk 24]
```diff
@@ -70,22 +80,19 @@ def _parse_args(self, *args, **kwargs):
                     self.ymin = args[0].ymin
                     self.deltay = args[0].deltay + offset
```
Comment (ismael-mendoza) [inline jax_galsim/bounds.py:81]: I think there is a weird edge case here. For this branch you just want to end up with the same value of `_isdefined` as the bounds that was passed in right? But I think in the case of `BoundsD` if `args[0]._isdefined == False` I think `self._isdefined` is going to end up `True` because of the logic: 

```python
if do_isdefined:
    self._isdefined = (self.deltax >= 0) & (self.deltay >= 0)
```

The simplest example to see this bug:

```python
assert not BoundsD()._isdefined
assert not BoundsD(BoundsD())._isdefined # crashes
```

For the fix, I think we just need to add `do_isdefined = False` in this branch also no?

File: jax_galsim/core/utils.py [hunk 25]
```diff
@@ -77,52 +77,61 @@ def cast_to_int(x, accept_strings=False):
...
-            return True
+                if kx not in y:
+                    currval &= arr_func(False)
+                else:
+                    currval &= is_equal_with_arrays(
+                        vx, y[kx], currval=currval, no_jax=no_jax
+                    )
         else:
-            return False
+            currval &= jnp.array(False)
```
Comment (ismael-mendoza) [inline jax_galsim/core/utils.py]: Why not `arr_func(False)` in this line?

File: jax_galsim/bounds.py [hunk 26]
```diff
@@ -708,3 +931,55 @@ def __hash__(self):
...
+        aux_data["deltax"] = self.deltax
+        aux_data["deltay"] = self.deltay
+        aux_data["isdefined"] = self._isdefined
+
+        return (children, aux_data)
+
+    @classmethod
+    def tree_unflatten(cls, aux_data, children):
+        """Recreates an instance of the class from flatten representation"""
+        ret = cls.__new__(cls)
```
Comment (ismael-mendoza) [inline jax_galsim/bounds.py:962]: Since you used `__new__` here you also have to set the `_dotypeconversion` attribute no?

File: jax_galsim/core/utils.py [hunk 27]
```diff
@@ -77,52 +77,61 @@ def cast_to_int(x, accept_strings=False):
     return _cast_to_type(x, int, accept_strings=accept_strings)
 
 
-def is_equal_with_arrays(x, y):
+def is_equal_with_arrays(x, y, currval=None, no_jax=False):
```
Comment (ismael-mendoza) [inline jax_galsim/core/utils.py]: Not a bug, but do you need to pass around the `currval` in this function? If any recursive branch returns `False` then it gets propagated automatically when it returns no? Might simplify the logic slightly

**Review by ismael-mendoza — CHANGES_REQUESTED** [review]
This looks pretty good Matt! I only found a couple of minor bugs and typos

File: jax_galsim/bounds.py (hunk 26 above)
Comment (beckermr) [reply jax_galsim/bounds.py:962]: done!

File: jax_galsim/core/utils.py (hunk 25 above)
Comment (beckermr) [reply jax_galsim/core/utils.py]: This was a bug. Good catch. Fixed!

File: jax_galsim/core/utils.py (hunk 27 above)
Comment (beckermr) [reply jax_galsim/core/utils.py]: Ack you are correct.

File: jax_galsim/bounds.py (hunk 24 above)
Comment (beckermr) [reply jax_galsim/bounds.py:81]: fixed!

**Comment by beckermr** [comment]
OK @ismael-mendoza. Thanks for the really good code review. This is ready for another look!

**Review by ismael-mendoza — APPROVED** [review]
Thanks for addressing the comments! Looks good  now!
